### ISSUE TITLE
⚠️Vortex DotNet Install Broken (Again)

### ISSUE THREAD

**Issue by Lii-Kun** [sonic2kk/steamtinkerlaunch#709]
## System Information
- SteamTinkerLaunch version: 12
- Distribution: POP!_OS 
- Installation Method: ProtonUpQt

## Issue Description

Installed Vortex with the menu option then started vortex. It started up fine but it has an red error box in the corner that says .net install broken and to install .net 4.6 or higher.

## Logs

[steamtinkerlaunch.log](https://github.com/sonic2kk/steamtinkerlaunch/files/10362332/steamtinkerlaunch.log)

**Comment by sonic2kk** [comment]
I recall you had issues with Vortex in #708. Did you definitely install Vortex into a clean Wineprefix?

Also, do you know if upstream Vortex had any updates recently that may have broken Wine compatibility?

**Comment by sonic2kk** [comment]
The attached log does not seem like it was from Vortex installation but from an attempt to launch Vortex. You should provide a log from the Vortex installation attempt.

Do you know if the DotNet6 executable is downloading and installing correctly? Perhaps the URL for it is outdated now.

**Comment by sonic2kk** [comment]
Just checked and the download link for dotnet6 looks valid.

Can you go to `~/.config/steamtinkerlaunch/vortex/downloads` and see if there is a `site` folder? If there is, check if it has a `windowsdesktop-runtime-6.0.6-win-x64.exe`, which should be about 55mb.

**Comment by Lii-Kun** [comment]
Yea it should be a clean prefix. It has been saying this from the first time I installed. Yea that file is in there.

**Comment by sonic2kk** [comment]
Interesting. Nothing should have broken on the SteamTinkerLaunch side so I suspect there's been a breaking change on the Vortex side. Do you know if anything has changed regarding Vortex and Wine compatibility recently?

I'm testing a fresh installation of Vortex now to see if it works on my end.

**Comment by sonic2kk** [comment]
I am able to reproduce the error:

![image](https://user-images.githubusercontent.com/7917345/211084569-82b40b36-3041-42b9-8631-4e0dc65b04e3.png)

No idea how to fix it yet, not even really sure what broke. Vortex doesn't even seem like it updated recently. Maybe dotnet6 installation is failing.

I don't use Vortex and don't know very much about the general tweaks used to fix it on Linux, so no ETA on when this will be fixed. In the meanwhile I recommend ModOrganizer 2, which *should* work.

I will keep this issue open to remind myself of the issue and perhaps to encourage others with more knowledge on Vortex on Linux to contribute :-)

**Comment by Lii-Kun** [comment]
The dotnet install log says: "warning: This package (dotnet48) is broken in wine-7.0. Broken since 5.18. See https://bugs.winehq.org/show_bug.cgi?id=49897 for more info. Use --force to try anyway." 
But going to that link it says it has been fixed.
[installDotNet.log](https://github.com/sonic2kk/steamtinkerlaunch/files/10362580/installDotNet.log)

Though you said dotnet6 is supposed to be installing but the log says only dotnet48
[steamtinkerlaunch.log](https://github.com/sonic2kk/steamtinkerlaunch/files/10362611/steamtinkerlaunch.log)

**Comment by sonic2kk** [comment]
That `installDotNet.log` is for Winetricks, which is used to install `dotnet48`. Not sure if this is strictly required anymore. The warning about it being broken can be ignored I think, but I'm not sure why it's showing up. Are you using an up-to-date Winetricks? You should be using `20220411-next`.

My installDotNet log shows a successful installation much more detailed than the one you are showing. I can't see any mention of your Winetricks version so just ensure you've updated it to the latest version.

`dotnet6` is installed via its installer (from the site folder) which is performed separately, as Winetricks does not have support for installing dotnet6 yet.

I don't believe there are any logs that could give a further hint. I am not sure why dotnet6 installation isn't enough, Vortex a few months ago actually updated to require dotnet6. Perhaps now it requires dotnet4X and dotnet6 is causing issues with that.

EDIT: Your Winetricks version is out of date, thanks for attaching the installation log. You are using `20210206`, you'll need to update which should get dotnet48 to install (hopefully). However I don't think that will fix the error that Vortex is giving. I will do a test installation of Vortex without installing dotnet6 and see if that fixes anything.

**Comment by sonic2kk** [comment]
Attempting to install Vortex without installing dotnet6 first results in an error that was fixed by adding the dotnet6 installation ages ago (Can't find the issue # now though).

![image](https://user-images.githubusercontent.com/7917345/211088958-f8488c9a-c391-46c1-956c-fc94d1c05505.png)

Pressing the "Fix" button here will install dotnet6 pretty much the same way STL does (except it uses the silent flag for the executable). It also results in the same error that dotnet >= 46 is required and that modding won't work. That rules out needing a newer dotnet6 executable link at least, if Vortex's own installation to fix it is failing then something must have changed. As of about a week ago this was [still working](https://github.com/sonic2kk/steamtinkerlaunch/issues/272#issuecomment-1368268440).

Unless something changed with Vortex in the background I'm not sure what is causing this. I'll pin the issue as it's fairly important, maybe someone can give a useful hint on what may be causing the problem.

Thanks for reporting this, sorry that I can't fix it. Such is the price of using .NET and Wine :frowning:

**Comment by sonic2kk** [comment]
Tested with a newer version of the dotnet6 desktop (6.0.12) by downloading it manually from `https://dotnet.microsoft.com/en-us/download/dotnet/6.0` and renaming it to the same filename that SteamTinkerLaunch expects. I removed the silent flag and the checks for dotnet6 already being installed but no dice, Vortex still gives the same error.

I tested with GE-Proton7-42 and GE-Proton7-43 and both worked. I also did some research online and found a handful of Windows users who had this issue and they "resolved" it by using MO2 :raised_eyebrow: 

Nothing has changed in the last few days with SteamTinkerLaunch at all, let alone for Vortex support, so this is something that has broken upstream. I don't know how to resolve it or what the full impact of this .NET problem is.

Maybe Vortex updated and broke wherever it looked for dotnet6 to be installed, and now it can't find it? It looks like dotnet6 has installed correctly as the installer seems to run correctly both with and without the silent flag, and when Vortex attempts to install it. So my earlier thought that this is a Wine issue may not be correct.

When running Vortex after yet another fresh installation, I noticed it said something about updating to a beta. I looked in the settings and saw that I was indeed running the beta. Rolling back seemed to go back really far, even though the latest release on GItHub is `1.7.3` this is apparently a beta. The "stable" is `1.6.XX`. Even rolling back gives the same warning, when it did not previously. So this seems like something has changed on the Vortex side and how it looks for dotnet, but I am not sure when this changed...

No leads at the moment I'm afraid. My money is still on upstream breaking Wine compatibility but I don't know where or how.

**Comment by Lii-Kun** [comment]
The error might just be caused by a bug in the way vortex checks for dotnet. I was able to install some mods in Fallout 4  even with the error.

**Comment by sonic2kk** [comment]
That is really positive news at least, I'm glad you were able to get some mods working! I also agree then that it is likely something Vortex has changed. I wonder where it thinks dotnet should be and if we could create any symlinks.

I couldn't find anything online about this other than some users saying Vortex couldn't find dotnet6 for them on Windows (same error but likely a different root cause). This sort of knowledge may have to come from someone else unless I stumble across some useful hint in my own travels.

**Comment by Lii-Kun** [comment]
While Vortex is putting the mod files into the game folders. They weren't enabled in game and in game the load order was different from what vortex was showing. It also didn't enable Archive Invalidation in the ini file. So something isn't working right.

**Comment by sonic2kk** [comment]
Ah, I guess that would be the result of missing DotNet6. Or rather, Vortex being unable to use it. Unfortunately I don't know what happened or how to fix it, though I hope it can be resolved soon

**Comment by sonic2kk** [comment]
Has there been any change with this upstream, or has anyone found a workaround for the problem (e.g. a manual way to tell Vortex where to find dotnet?)

**Comment by sonic2kk** [comment]
Spent a bit of time looking into this and I can't find out what might have broken this. I recall the `DOTNET_ROOT` variable was implemented as a result of #555, but I tried fiddling around with that variable and setting various different paths to no avail. Vortex always complains about a missing or broken .NET installation.

**Comment by sonic2kk** [comment]
I saw a comment earlier today that this may be caused by bad registry keys (even on Windows), but I did a quick check and everything looked fine, the paths the keys pointed to looked like they existed.

**Comment by AutonomousMidnight** [comment]
> ![image](https://user-images.githubusercontent.com/7917345/211088958-f8488c9a-c391-46c1-956c-fc94d1c05505.png)

I thought that I should mention that the ".NET Installation Broken" message in Vortex notifications has been the norm with me for months. Though, I'm still able to install mods but I do have to create symlinks so Vortex can properly deploy and enable mods. It sucks, but it works none the less.

Since yesterday though, Vortex broke (haven't touched it for over a month) and it displays the message that I quoted l, only allowing me to either Quit Vortex or Fix the Dotnet installation, which never does anything since it is always looking in the root directory of my linux installation for dotnet.

I found ways around this by using winecfg to set the drive linking my linux root directory to something like /home/{youruser}. Seems now that everything is back to normal, though I still get the .NET Installation Broken message. At least I'm able to install mods now.

Sorry if this post is a little confusing as I'm pretty new to Linux and STL. I just thought that maybe if I give some amateur experience, it could provide a "GREAT SCOTT" moment for someone here.

**Comment by sonic2kk** [comment]
> it is always looking in the root directory of my linux installation for dotnet.

Huh, this is interesting! I wonder if Vortex is looking for a Linux-native DotNet ? 

Could you try installing dotnet core native e.g. from your package manager? Maybe Vortex is expecting a native Linux DotNet now (though this isn't documented anywhere afaik).

If that works once I have a bit more time I can experiment and see if any of the Winetricks are required.

Otherwise, the workaround you mentioned could potentially be implemented by SteamTinkerLaunch. Though I'd prefer to not create files like that if I can help it.

<hr>

It's interesting that you note that it's been broken for a while. I hadn't reinstalled Vortex on my PC for a long time as I don't use Vortex and only had it installed for fixing some of the Vortex command line functions.

<hr>

The Vortex installation stuff wasn't really touched in the last month, so I'm fairly sure that this wasn't broken by STL but that there were upstream changes that broke things on Linux.

Thanks for the details and background, that's extremely valuable! 😄

**Comment by AutonomousMidnight** [comment]
> Could you try installing dotnet core native e.g. from your package manager? Maybe Vortex is expecting a native Linux DotNet now (though this isn't documented anywhere afaik).

I should of been more descriptive here as there are a few things I didn't mention. I do have the native linux dotnet installed and Vortex detects the exact folder where dotnet is located. The problem is Vortex wants the Windows binaries of dotnet, which are missing.

I started to add in files that are for Windows, such as hostfxr.dll to see if there are any changes, but then it complains about a new file that is missing. (I didn't replace any files as I don't want to break my linux dotnet installation). I could continue this for testing purposes.

**Comment by sonic2kk** [comment]
Not sure if it's worth keeping this open. This seems to be a Vortex issue but there are users that are seemingly using fresh Vortex installs with STL, and likely no progress will be made on this issue.

I'll see if I can still reproduce this later and take it from there :-)

EDIT: There was an update available for Vortex, but the error is still there even after updating. A reinstall could fix it but I will test that another time.

**Comment by zany130** [comment]
fwiw I was able to get vortex running for Hogwarts legacy today on my PC (I couldn't get it working on my steam deck) but vortex doesn't detect Hogwarts legacy. Going into the game menu within the extra vortex menu does not show Hogwarts as a game to enable vortex for.

Don't know if I doing something wrong never used mod manager like vortex before but my understanding is that steamtinkerlaunch is checks your installed steam games and comparing it to a list of vortex supported games and symlinking them into vortex's wine prefix?

Hogwarts legacy doesn't seem to be natively supported as you have to install a vortex extension for it (again may be wrong never used vortex before but it needed that extension on windows and I am assuming most games don't need extensions?)

So I am guessing (again abosulety no experience with this first time doing this) steamtinkerlaunch is not symlinking hogwarts because it's not natively compatible?

**Comment by sonic2kk** [comment]
> fwiw I was able to get vortex running for Hogwarts legacy today on my PC (I couldn't get it working on my steam deck)

Thank you for testing! Interesting that it didn't work on Steam Deck. I don't use Vortex and on top of that I don't like using mods on my Steam Deck in almost every instance, so I have never even attempted to use Vortex on Steam Deck. Good to know that you didn't get that dotnet issue. I will try a re-install in the near future and see if Vortex installs correctly :-) 

> Don't know if I doing something wrong never used mod manager like vortex before but my understanding is that steamtinkerlaunch is checks your installed steam games and comparing it to a list of vortex supported games and symlinking them into vortex's wine prefix?

That is pretty much my understanding too :-)

> So I am guessing (again abosulety no experience with this first time doing this) steamtinkerlaunch is not symlinking hogwarts because it's not natively compatible?

 I don't know if the extension thing is the issue. I think this is really because Hogwarts Legacy is not in the [`vortexgames.txt`](https://github.com/sonic2kk/steamtinkerlaunch/blob/master/misc/vortexgames.txt) file. I can get a PR up to add it though and it *should* be detected, though I'm not sure. I have heard mixed reports of games being detected this way. Perhaps games that need extensions won't be detected at all.

The entries in this file are formatted as `"domainname";"Vortex Name";"SteamAppID"`:
- `domainname` is what Vortex sees the game as. This is its filename in the Vortex web API and also the Vortex internal name, which can be seen from the unpacked Vortex Electron app in its prefix. In this case it could be `hogwartslegacy` but I didn't check.
- `Vortex Name` is a "pretty" name that Vortex calls the game as, typically mirroring the name used on the mod page. In this case, it might be `Hogwarts Legacy`
- Steam AppID, as we **only** support games purchased from Steam. In this case Hogwarts Legacy's AppID is `990080`

I can get a PR up to add Hogwarts Legacy but first I would probably check if other games need to be added, and do it as part of one slightly bigger PR :smile: 

> it needed that extension on windows and I am assuming most games don't need extensions?

Yup, some games need extensions. The Yakuza series needs RyuModManager as an extension. These extensions tend to be community mod managers that Vortex just downloads and runs most of the time, or passes information to.

<hr>

Given that I don't use Vortex, anything outside of critical Vortex breakages on STL's side is not really something I have kept up with too much. I fixed a lot of flaky command-related stuff for v12.0 and that's about it. You could try manually adding Hogwarts Legacy and seeing if that detects it.

**Comment by zany130** [comment]
Yeah, so far, I have gotten everything working by doing the following steps.

1. Add hogwarts legacy to the vortex steam collection
2. enable vortex in the game menu (I'm guessing this is redundant because of the above)
3. start the game
4. open vortex when I get the prompt
5. go to games and search and select Hogwarts
6. it won't auto-detect the path, so I have to manually specify the path (hint make a symlink on your desktop to the game folder to make it easier to select the folder from the wine browser)
7. Vortex will prompt you to install the extension, so install it
8. vortex will need to be reset, unfortunately, this makes STL think your done moding and will start the game so you need to exit the game
9. install your mods by downloading from nexus mod website and installing them with the install from file in the mods page
10. mods are installed into the vortex prefix (under `~/.config/steamtinkerlaunch/vortex/compatdata/pfx/drive_c/users/steamuser/AppData/Roaming/Vortex/hogwartslegacy/mods/`) so you need to symlink them into your game folder ( for Hogwarts this is `~/.local/share/Steam/steamapps/common/Hogwarts Legacy/Phoenix/Content/Paks/~mods/`)
11. enable the mods you want
12. exit vortex, and your mods should load


FWIW, I used proton-ge 7.49. Maybe that fixed vortex?

**Comment by sonic2kk** [comment]
Thanks for outlining the steps. I am pretty sure this is simply because SteamTinkerLaunch is not configured to see Hogwarts Legacy. That file is **not automatically updated** and is **not intended to be**, it's intended to be manually updated to add games. Hogwarts Legacy is not in that file, so SteamTinkerLaunch is not doing anything wrong here, it's actually working as intended.

Once I have some time I can look into adding this game and some others, though really support for this file is meant to come from active Vortex users, not you or me who are not very experienced with Vortex. Still, I can look into adding it. Based on the `Roaming/Vortex` path you gave, I think that solves the biggest pain about adding games which is finding out the Vortex domainname.

> FWIW, I used proton-ge 7.49. Maybe that fixed vortex?

Thanks for the hint, I will try using that when I reinstall (probably not tonight, but will make a note for future testing :-))

**Comment by sonic2kk** [comment]
Added the new games I could see that were added recently to Vortex with 968e1854cbe891c943cab08f6ee1224314c9f614

**Comment by sonic2kk** [comment]
Uninstalling by removing `~/.config/steamtinkerlaunch/vortex` and re-installing by first selecting the Vortex Proton version in the Global Menu as "GE-Proton7-49" and then running `steamtinkerlaunch vortex start` fixes the issue for me. Vortex did hang initially at "loading mod content" or something like that, and I had to force kill the Wine and Chromium background processes, but that might've been a one-off hiccup.

Closing this issue as it appears to be resolved now with newer community Proton versions. Thanks @zany130 for the hint! I am not sure why it wasn't working on Steam Deck but it can be investigated separately (I suspect the dotnet installation is failing).
